Composer/packages/client/src/pages/botProject/skill-configuration/AllowedCallers.tsx

```diff
@@ -0,0 +1,170 @@
+  marginTop: '4px',
+});
+
+const Row = styled(Stack)({
+  borderBottom: `1px solid ${FluentTheme.palette.neutralLight}`,
+  padding: '8px 0 8px 4px',
+  '& .ms-Button:not(:focus) i': {
+    visibility: 'hidden',
+  },
+  '&:hover .ms-Button i': {
+    visibility: 'visible',
+  },
+});
+
+type ItemProps = {
+  value: string;
+  onBlur: () => void;
+  onChange: (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue: string) => void;
+  onRemove: () => void;
+};
+
+const Item = React.memo(({ value, onBlur, onChange, onRemove }: ItemProps) => {
+  const itemRef = React.useRef<ITextField | null>(null);
+  const didMount = React.useRef<boolean>(false);
+
+  React.useEffect(() => {
+    if (!value && !didMount.current) {
+      itemRef.current?.focus();
+    }
+    didMount.current = true;
+  }, []);
+
+  return (
+    <Row horizontal>
+      <Input
+        componentRef={(ref) => (itemRef.current = ref)}
+        styles={textFieldStyles}
+        value={value}
+        onBlur={onBlur}
+        onChange={onChange}
+      />
+      <IconButton aria-label={formatMessage('Remove item')} iconProps={{ iconName: 'Trash' }} onClick={onRemove} />
+    </Row>
+  );
+});
+
+type Props = {
+  projectId: string;
+};
+
+export const AllowedCallers: React.FC<Props> = ({ projectId }) => {
+  const { setSettings } = useRecoilValue(dispatcherState);
+  const rootBotProjectId = useRecoilValue(rootBotProjectIdSelector) || '';
+  const settings = useRecoilValue(settingsState(projectId));
+  const mergedSettings = mergePropertiesManagedByRootBot(projectId, rootBotProjectId, settings);
+  const { skillConfiguration } = mergedSettings;
+
+  const setAllowedCallers = React.useCallback(
+    (allowedCallers: string[] = []) => {
+      const updatedSetting = {
+        ...cloneDeep(mergedSettings),
+        skillConfiguration: { ...skillConfiguration, allowedCallers },
+      };
+      setSettings(projectId, updatedSetting);
+    },
+    [mergedSettings, projectId, skillConfiguration]
+  );
+
+  const onBlur = React.useCallback(() => {
+    const updatedAllowedCallers = [...skillConfiguration?.allowedCallers];
+    setAllowedCallers(updatedAllowedCallers.filter(Boolean));
+  }, [skillConfiguration?.allowedCallers, setAllowedCallers]);
+
+  const onChange = React.useCallback(
+    (index: number) => (_: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue = '') => {
+      const updatedAllowedCallers = [...skillConfiguration?.allowedCallers];
+      updatedAllowedCallers[index] = newValue;
+      setAllowedCallers(updatedAllowedCallers);
+    },
+    [skillConfiguration?.allowedCallers, setAllowedCallers]
+  );
+
+  const onRemove = React.useCallback(
+    (index: number) => () => {
+      const updatedAllowedCallers = skillConfiguration?.allowedCallers?.filter((_, itemIndex) => itemIndex !== index);
+      setAllowedCallers(updatedAllowedCallers);
+    },
+    [skillConfiguration?.allowedCallers, setAllowedCallers]
```

Comment:
no need for set... as dep

Reply:
Same as above. `setAllowedCallers` can update so it should be a required dep.